### PR TITLE
Fix multiple networks with separately specified ip and mac

### LINKED ISSUE #817: With multiple networks: Unable to specify per-network static ip addresses
When attempting to run a service with multiple networks, each with a static IP address, podman is incorrectly invoked with multiple `--ip` parameters.

It should instead use `--network <name>:ip=<ip>`. This problem might possibly extend to other per-network options, such as mac addresses to use.

**To Reproduce**

Example `compose.yml`:

```yaml
version: "3.8"
services:
  hello-world:
    image: docker.io/hello-world:latest
    tty: true
    stdin_open: true
    networks:
      net0:
      net1:
        ipv4_address: 192.168.72.2
      net2:
        ipv4_address: 192.168.72.34

networks:
  net0:
    driver: bridge

  net1:
    driver: ipvlan
    ipam:
      config:
        - subnet: 192.168.72.0/27

  net2:
    driver: ipvlan
    ipam:
      config:
        - subnet: 192.168.72.32/27
```

**Expected behavior**

`podman create` succeeds as it is invoked with `--net podman-compose-net-bug_net0 --net podman-compose-net-bug_net1:ip=192.168.72.2 --net podman-compose-net-bug_net2:ip=192.168.72.34` 

**Actual behavior**

`podman create` fails because instead `--ip` is provided multiple times, but is not associated to the relevant network.

**Output**

```
$ podman-compose version

podman-compose version: 1.0.6
['podman', '--version', '']
using podman version: 4.7.2
podman-compose version 1.0.6
podman --version 
podman version 4.7.2
exit code: 0

$ podman-compose up

podman-compose version: 1.0.6
['podman', '--version', '']
using podman version: 4.7.2
** excluding:  set()
['podman', 'ps', '--filter', 'label=io.podman.compose.project=podman-compose-net-bug', '-a', '--format', '{{ index .Labels "io.podman.compose.config-hash"}}']
['podman', 'network', 'exists', 'podman-compose-net-bug_net0']
['podman', 'network', 'create', '--label', 'io.podman.compose.project=podman-compose-net-bug', '--label', 'com.docker.compose.project=podman-compose-net-bug', '--driver', 'bridge', 'podman-compose-net-bug_net0']
['podman', 'network', 'exists', 'podman-compose-net-bug_net0']
['podman', 'network', 'exists', 'podman-compose-net-bug_net1']
['podman', 'network', 'create', '--label', 'io.podman.compose.project=podman-compose-net-bug', '--label', 'com.docker.compose.project=podman-compose-net-bug', '--driver', 'ipvlan', '--subnet', '192.168.72.0/27',
 '--gateway', '192.168.72.1', 'podman-compose-net-bug_net1']
['podman', 'network', 'exists', 'podman-compose-net-bug_net1']
['podman', 'network', 'exists', 'podman-compose-net-bug_net2']
['podman', 'network', 'create', '--label', 'io.podman.compose.project=podman-compose-net-bug', '--label', 'com.docker.compose.project=podman-compose-net-bug', '--driver', 'ipvlan', '--subnet', '192.168.72.32/27'
, '--gateway', '192.168.72.33', 'podman-compose-net-bug_net2']
['podman', 'network', 'exists', 'podman-compose-net-bug_net2']
podman create --name=podman-compose-net-bug_hello-world_1 --label io.podman.compose.config-hash=7545ab99bc6998d28cd4d6061a44cada39f13dbe3a3643f49abfe5d82f56d34d --label io.podman.compose.project=podman-compose-n
et-bug --label io.podman.compose.version=1.0.6 --label PODMAN_SYSTEMD_UNIT=podman-compose@podman-compose-net-bug.service --label com.docker.compose.project=podman-compose-net-bug --label com.docker.compose.proje
ct.working_dir=/home/matteo/podman-compose-net-bug --label com.docker.compose.project.config_files=compose.yml --label com.docker.compose.container-number=1 --label com.docker.compose.service=hello-world --net p
odman-compose-net-bug_net2,podman-compose-net-bug_net1,podman-compose-net-bug_net0 --network-alias hello-world --ip=192.168.72.2 -i --tty docker.io/hello-world:latest
Error: --ip can only be set for a single network: invalid argument
exit code: 125
podman start -a podman-compose-net-bug_hello-world_1
Error: no container with name or ID "podman-compose-net-bug_hello-world_1" found: no such container
exit code: 125
```

**Environment:**
 - OS: Linux
 - podman version: 4.7.2 
 - podman compose version: 1.0.6

### PR DESCRIPTION
**Pull request by baszoetekouw**
Currently, podman-compose incorrectly parses docker-compose file with multiple networks, for which the `ipv4_address`, `ipv6_address` or `mac_address` is specified for each network (see eg #817).  There is a partial fix in devel, which solves this issue for the case that the networks addresses are explicitly specified for all defined addresses.  However, two cases still fail:
 - if the `mac_address` is specified (either globally or per interface)
 - if one interface has no explicit ipv4 address, and one does;  this happens for example when defining an internal net (which should use dynamic addresses) and a macvlan interface with a fixed ip on the host network.

To be more specific, current HEAD of podman-compose, using this configuration:
```yaml
---
version: "3"
services:
  number_1:
    image: busybox
    networks:
      internal:
  number_2:
    image: busybox
    networks:
      external:
        ipv4_address: "192.168.1.242"
        mac_address: "02:01:42:00:01:03"
      internal:

networks:
  internal:
    driver: "bridge"
    internal: true
    dns_enabled: true
  external:
    driver: "macvlan"
    driver_opts:
      parent: "eth0"
    ipam:
      driver: "host-local"
      config:
        - subnet: "192.168.1.0/24
```

fails with: 
```
podman-compose version: 1.0.7
['podman', '--version', '']
using podman version: 4.3.1
** excluding:  set()
['podman', 'ps', '--filter', 'label=io.podman.compose.project=tmp', '-a', '--format', '{{ index .Labels "io.podman.compose.config-hash"}}']
podman pod create --name=pod_tmp --infra=false --share=
49e035ad3839ef2d6a6a4d4c1174fb0e0e7fccf3f9dcedfc9b8dd3edcd8f5bfc
exit code: 0
['podman', 'network', 'exists', 'tmp_internal']
podman create --name=tmp_number_1_1 --pod=pod_tmp --label io.podman.compose.config-hash=eef2914b01b6c8cee877f8ffb53431be4086f78a986ece813948eac9bf7d93ef --label io.podman.compose.project=tmp --label io.podman.compose.version=1.0.7 --label PODMAN_SYSTEMD_UNIT=podman-compose@tmp.service --label com.docker.compose.project=tmp --label com.docker.compose.project.working_dir=/tmp --label com.docker.compose.project.config_files=test1.yml --label com.docker.compose.container-number=1 --label com.docker.compose.service=number_1 --net tmp_internal --network-alias number_1 busybox
a35b3df3e0497580f9b6d9ca3d07026200e2b477e7b63f12d048210c31702fac
exit code: 0
['podman', 'network', 'exists', 'tmp_external']
['podman', 'network', 'exists', 'tmp_internal']
podman create --name=tmp_number_2_1 --pod=pod_tmp --label io.podman.compose.config-hash=eef2914b01b6c8cee877f8ffb53431be4086f78a986ece813948eac9bf7d93ef --label io.podman.compose.project=tmp --label io.podman.compose.version=1.0.7 --label PODMAN_SYSTEMD_UNIT=podman-compose@tmp.service --label com.docker.compose.project=tmp --label com.docker.compose.project.working_dir=/tmp --label com.docker.compose.project.config_files=test1.yml --label com.docker.compose.container-number=1 --label com.docker.compose.service=number_2 --net tmp_internal,tmp_external --network-alias number_2 --ip=192.168.1.254 busybox
Error: --ip can only be set for a single network: invalid argument
exit code: 125
[number_1] podman start -a tmp_number_1_1
[number_2] podman start -a tmp_number_2_1
[number_2] | Error: no container with name or ID "tmp_number_2_1" found: no such container
[number_2] exit code: 125
[number_1] exit code: 0
```

This is caused by podman-compose parsing the compose file network config to 
```
--net tmp_internal,tmp_external --ip=192.168.1.254
```
which in invalid.

With this PR, it executes
```
podman create --name=tmp_number_2_1 --pod=pod_tmp --label io.podman.compose.config-hash=eef2914b01b6c8cee877f8ffb53431be4086f78a986ece813948eac9bf7d93ef --label io.podman.compose.project=tmp --label io.podman.compose.version=1.0.7 --label PODMAN_SYSTEMD_UNIT=podman-compose@tmp.service --label com.docker.compose.project=tmp --label com.docker.compose.project.working_dir=/tmp --label com.docker.compose.project.config_files=test1.yml --label com.docker.compose.container-number=1 --label com.docker.compose.service=number_2 --network tmp_external:ip=192.168.1.254,mac=02:01:42:00:01:03 --network tmp_internal busybox
```
so the network arguments are 
```
--network tmp_external:ip=192.168.1.254,mac=02:01:42:00:01:03 --network tmp_internal
```
instead.

Note that I've added a test case for this issue in [tests/test_podman_compose_networks.py](https://github.com/containers/podman-compose/compare/devel...baszoetekouw:podman-compose:fix-networks?expand=1#diff-6428e43e2d7ab25a856bb9a4b6e4b194eca84b8990a04625eb29281b25549c09).  I'm not entirely sure this this is the correct way to integrate the tests, so please let me know if I should adjust that.

Closes #817

### CODE REVIEW

**Review by p12tic — CHANGES_REQUESTED** [review]
Looks good in principle.

The test suite has moved to unittest in main branch, so this PR needs to be updated to match.

Also I think it would be useful to add explicit unit tests to get_net_args function. See e.g. what I did in this PR: https://github.com/containers/podman-compose/pull/819/files. 

With unit tests we can cover way more special cases as they run so fast.

**Comment by baszoetekouw** [comment]
Hi @p12tic !  I'm working on the unit tests as you suggested, but I encountered a bit of a problem.

It turns out that docker-compose doesn't support specifying a `mac_address` for a specific network interface.  Instead, it only allows specification of `mac_address` per container, and seems to randomly apply it to one or more of the specified network interfaces (haven't tried this myself as I don't currently have access to a machine running docker, but this seems to be the consensus on stackexchange).

Of course, podman is much nicer than docker and properly supports specification of `mac_address`per network (with similar semantics as for ipv4 and ipv6 addresses), so it would be nice if podman-compose could also support this.  

There are two choices to make here.  First, how to handle the container-level `mac_address` option.  I see three options:
   1. apply the `mac_address` the to all specified networks in the container
   2. apply the `mac_address` to the first network interface specified
   3. apply the `mac_address` to the first macvlan network interface specified

I would expect that option 1 might lead to network headaches inside the container, that option 2 is probably most like the original docker-compose behaviour, and option 3 might mostly do what people might expect (as there is usually no reason to specify `mac_address` unless you're using a macvlan interface).

And then there's the question if you would like podman-compose to support "extended" podman functionality and allow specification of the `mac_addreess` per interface, like this:
```yaml
services:
    some_service:
        image: busybox
        hostname: myhost
        command: ["/bin/busybox", "httpd", "-f", "-h", "/var/www/html", "-p", "8001"]
        working_dir: /var/www/html
        networks:
            shared-network:
                ipv4_address: "172.19.1.10"
                mac_address: "02:01:01:00:01:01"
            internal-network:
                ipv4_address: "172.19.2.10"
                mac_address: "02:01:01:00:02:01"
```

Please let me know how you would like me to implement this.

**Comment by p12tic** [comment]
docker-compose spec says it just passes the mac address as `--mac-address` option to `docker run`. In the docker CLI code there's this logic:
 - if there's per-network mac-address and container mac address, exit with error
 - otherwise, apply mac address to the attachment to the first network if there are multiple networks, or to the attachment to the default network. 

I think it makes sense to preserve this behavior for the container level `mac_address`. 

As for network-level `mac_address`, we can add this option as `podman.mac_address`. If you end up adding it, then please  also document it as a podman-compose specific extension in a new file in `docs/`.

**Comment by baszoetekouw** [comment]
I've made the changes as you requested:
 - mac_address support is now identical to docker-composer; the podman-specific per-network `mac_address` settings are now called `podman.mac_address` and this extension is documented in docs/Extensions.md
 - I've added unit tests as you suggested.  As a bonus, `get_net_args()` now has 100% coverage
 - I've fixed a few bugs in the handling of the `network_mode` parameter that I encountered while writing the unittests
 - I've added the last missing podman-specific `network_mode` values (`private` and `pasta`), and documented those (together with the pre-existing podman-specific values `slirp4netns` and `ns:`) also in docs/Extensions.md
 - synced with updated test suites in main

File: pytests/test_get_net_args.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+            f"--network={PROJECT_NAME}_net1:ip={ip4_1}",
+            f"--network={PROJECT_NAME}_net2:ip={ip6_2}",
+            f"--network={PROJECT_NAME}_net3",
+            f"--network-alias={SERVICE_NAME}",
+        ]
+        args = get_net_args(compose, container)
+        self.assertListEqual(expected_args, args)
+
+    @parameterized.expand([
+        # TODO: check is this behaviour introduced by this patch?
```
Comment (p12tic) [inline pytests/test_get_net_args.py:259]: Leftover that probably shouldn't be in a PR?

**Review by p12tic — CHANGES_REQUESTED** [review]
Just one small nit and we can merge this in. Thanks a lot!

**Comment by baszoetekouw** [comment]
Indeed, I forgot to remove that comment.  Fixed now!

**Pull request merged**